pkg/objstore/objstore.go

```diff
@@ -137,12 +140,36 @@ func TryToGetSize(r io.Reader) (int64, error) {
 		return fileInfo.Size(), nil
 	case *bytes.Buffer:
 		return int64(f.Len()), nil
+	case *bytes.Reader:
+		// Returns length of unread data only.
+		return int64(f.Len()), nil
 	case *strings.Reader:
 		return f.Size(), nil
+	case Sizer:
+		return f.Size()
 	}
 	return 0, errors.Errorf("unsupported type of io.Reader: %T", r)
 }
 
+// Sizer can return size of data.
+type Sizer interface {
+	// Size returns length of the data, or error, if it is not available.
+	// It is best to call Size before calling other methods on the type, as some io.Reader implementations may only return
```

Comment:
I see how tricky it is now. If we assume `Size` is unread data then all our implementations have to be much more complex for no reason, as Upload only care about the overall initial size before reading 🙃 

Since we are in `objstore.Sizer`, maybe `ObjectSize` method will be easier to understand? I would focus on capabilities we want which is overall size. It will much more obvious. WDYT?

Reply:
> I see how tricky it is now. If we assume Size is unread data then all our implementations have to be much more complex for no reason, as Upload only care about the overall initial size before reading 🙃

What I want to express is that *callers* should call `Size()` before doing any other operations because *implementations* may be able to only answer `Size` with **remaining data** (like `(*bytes.Reader).Len()` method)

`Upload` is already doing that – calls `Size` before doing anything else, and requires no change.

Reply:
Right. That's why I would propose naming this method `ObjectSize` to reflect the initial object size, no matter when you do this. 

But I guess it might be too picky. I would be fine to leave it as it is in your case. Happy with that (:

Reply:
> That's why I would propose naming this method ObjectSize to reflect the initial object size, no matter when you do this.

Sure, we can rename it. We can also specify that Sizer must report full object size. 

What do you say if we rename `Sizer` to `ObjectSizer`, while at it?